andi/andi.py

```diff
@@ -53,10 +59,161 @@ def to_provide(
+                                dependency injection for these classes where
+                                we don't want it because they will be provided by
+                                other means.
+    :return: The plan ready to be used as ``build`` method input.
+    """
+    assert can_provide is not None
+    assert externally_provided is not None
+    if isinstance(can_provide, Container):
+        can_provide = can_provide.__contains__
+    if isinstance(externally_provided, Container):
+        externally_provided = externally_provided.__contains__
+    bindings = bindings or (lambda x: None)
+    if isinstance(class_or_func, type):
+        # This covers applying binding to input class itself
+        class_or_func = bindings(class_or_func) or class_or_func
+    return _plan(class_or_func, can_provide, externally_provided, bindings, None)
+
+
+def _plan(class_or_func: Union[Type, Callable],
+          can_provide: Callable[[Optional[Type]], bool],
+          externally_provided: Callable[[Optional[Type]], bool],
+          bindings: Callable[[Type], Optional[Type]],
+          dependency_stack=None) -> Plan:
+    dependency_stack = dependency_stack or []
+    plan_seq = OrderedDict()  # type: Plan
+    type_for_arg = {}
+
+    input_is_type = isinstance(class_or_func, type)
+
+    if input_is_type:
+        cls = typing.cast(Type, class_or_func)
+        if not can_provide(cls):
+            raise NonProvidableError(
+                "Type {} cannot be provided".format(as_class_names(cls)))
+
+        if externally_provided(cls):
+            plan_seq[cls] = {}
+            return plan_seq
+
+        if cls in dependency_stack:
+            raise CyclicDependencyError(
+                "Cyclic dependency found. Dependency graph: {}".format(
+                    " -> ".join(as_class_names(dependency_stack + [cls]))))
+        dependency_stack = dependency_stack + [cls]
+        arguments = inspect(cls.__init__)
+    else:
+        arguments = inspect(class_or_func)
+
+    for argname, types in arguments.items():
+        sel_cls = select_type(types, can_provide, bindings)
+        if sel_cls is not None:
+            if sel_cls not in plan_seq:
+                plan_seq.update(_plan(sel_cls, can_provide, externally_provided,
+                                      bindings, dependency_stack))
+            type_for_arg[argname] = sel_cls
+        else:
+            # Non fulfilling all deps is allowed for non type inputs.
+            if input_is_type:
+                msg = "Any of {} types are required ".format(as_class_names(types))
+                msg += " in {} __init__ but none can be provided".format(
+                        as_class_names(class_or_func))
+                raise NonProvidableError(msg)
+
+    plan_seq[cls if input_is_type else FunctionArguments] = type_for_arg
+    return plan_seq
+
+
+def build(plan: Plan, stock: Optional[Dict[Type, Any]] = None):
```

Comment:
Would it be ok if we discuss `build` function separately, and merge this PR without it? It requires docs and README update, and it seems it is not needed for scrapy-po.

Reply:
I see one use of this function: it is documented that andi.plan returns something which can be passed to andi.build. It seems understanding andi.build can be important for understanding of andi.plan. What do you think about either moving andi.build to README, or copy-pasting it to README?

Reply:
This function is also useful for andi unit testing, as it helps to test that the andi plans are actually executable. What do you think about moving it to test utils? I think it is also worth to have it in the README as an example to see how to use a generated plan. 

> What do you think about either moving andi.build to README, or copy-pasting it to README?

Are not both the same thing?

Reply:
> This function is also useful for andi unit testing, as it helps to test that the andi plans are actually executable. What do you think about moving it to test utils? I think it is also worth to have it in the README as an example to see how to use a generated plan.

👍

Reply:
> Are not both the same thing?

Sorry for the unclear writing :) I wondered if it should be kept only in readme, or in readme and source code.

Reply:
I have include examples in the `plan_for_func` and `plan_for_class` in https://github.com/scrapinghub/andi/pull/8 . I will plan to also update the README once we have clear the API